grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java

```diff
@@ -0,0 +1,203 @@
+ * Encapsulates the state of a single server call, reading messages from the client, passing to business logic
+ * via {@link Listener}, and writing messages passed back to the response.
+ */
+final class UnaryServerCall<I, O> extends AbstractServerCall<I, O> {
+
+    private final HttpRequest req;
+    private final CompletableFuture<HttpResponse> resFuture;
+    private final ServiceRequestContext ctx;
+    private final UnaryMessageDeframer requestDeframer;
+
+    // Only set once.
+    @Nullable
+    private O responseMessage;
+    private boolean deframingStarted;
+
+    UnaryServerCall(HttpRequest req,
+                    MethodDescriptor<I, O> method,
+                    String simpleMethodName,
+                    CompressorRegistry compressorRegistry,
+                    DecompressorRegistry decompressorRegistry,
+                    // TODO(ikhoon): Add CompletableFuture<HttpResponse>
+                    HttpResponse res,
+                    CompletableFuture<HttpResponse> resFuture,
+                    int maxRequestMessageLength,
+                    int maxResponseMessageLength,
+                    ServiceRequestContext ctx,
+                    SerializationFormat serializationFormat,
+                    @Nullable GrpcJsonMarshaller jsonMarshaller,
+                    boolean unsafeWrapRequestBuffers,
+                    boolean useBlockingTaskExecutor,
+                    ResponseHeaders defaultHeaders,
+                    @Nullable GrpcStatusFunction statusFunction) {
+        super(req, method, simpleMethodName, compressorRegistry, decompressorRegistry, res,
+              maxResponseMessageLength, ctx, serializationFormat, jsonMarshaller, unsafeWrapRequestBuffers,
+              useBlockingTaskExecutor, defaultHeaders, statusFunction);
+        requireNonNull(req, "req");
+        this.ctx = requireNonNull(ctx, "ctx");
+        final boolean grpcWebText = GrpcSerializationFormats.isGrpcWebText(serializationFormat);
+        requireNonNull(decompressorRegistry, "decompressorRegistry");
+
+        final RequestHeaders clientHeaders = req.headers();
+        requestDeframer = new UnaryMessageDeframer(ctx.alloc(), maxRequestMessageLength, grpcWebText)
+                .decompressor(clientDecompressor(clientHeaders, decompressorRegistry));
+        this.req = req;
+        this.resFuture = requireNonNull(resFuture, "resFuture");
+    }
+
+    @Override
+    public void request(int numMessages) {
+        if (ctx.eventLoop().inEventLoop()) {
+            request0();
+        } else {
+            ctx.eventLoop().execute(this::request0);
+        }
+    }
+
+    private void request0() {
+        if (listener() == null) {
+            return;
+        }
+        startDeframing();
+    }
+
+    @Override
+    void startDeframing() {
+        if (deframingStarted) {
+            return;
+        }
+        deframingStarted = true;
+        req.aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc()).handle((agg, cause) -> {
```

Comment:
Would it be better collecting the received `HttpData` instances as they are rather than merging them? If we received only one `HttpData` (excluding the empty ones), we could use `UnaryMessageDeframer` but otherwise we could use less optimized one.